### ISSUE TITLE
How to fix "Connection pool is full, discarding connection" warning?

### ISSUE THREAD

**Issue by alex4200** [locustio/locust#1647]
### Describe the bug

When I make more than 15 requests on locust in parallel, I get the warning

`WARNING/urllib3.connectionpool: Connection pool is full, discarding connection:`

which, according to a comment [HERE](https://github.com/oliver-zehentleitner/unicorn-binance-websocket-api/issues/80), is caused by "logging"(?), and could be the reason of some underlying issue. 

I tried to use the line `logging.basicConfig(level=logging.ERROR)` in the `locustfile.py`, but this does not solve the issue. 

### Expected behavior

The expected behavior is that locust is able to perform, like, 50 requests in parallel for 50 users. That is what I would expect from a performance testing tool. 

### Actual behavior

I get the warning 

`WARNING/urllib3.connectionpool: Connection pool is full, discarding connection:`

and I am not sure locust is doing the correct thing (i.e. making like 20 requests in parallel). 

### Steps to reproduce

Here is the part of the code that is (hopefully) doing a bunch of requests in parallel (just like a browser would do is a user opens a web page):

```
    def make_parallel_requests(self, requests):
        group = Group()
        uid = str(uuid.uuid4())[:8]
        print(f"{uid} Starting {len(requests)} requests in parallel.")
        for request in requests:
            if request["method"] == "GET":
                url = request["url"]
                headers = request["headers"]
                headers["Authorization"] = "bearer " + ACCESS_TOKEN
                group.spawn(lambda: self.client.get(url, headers=headers, verify=False))
        group.join()
        print(f"{uid} Requests done")
```

It looks like that if the number of requests is smaller than 15 there is no such warning. But there will be printed that warning for every additional request. 

Maybe I can ignore the warning and locust is, in fact, performing the requests in parallel? 

### Environment

- OS: MacOS 10.14.6
- Python version: 3.8.6
- Locust version: 1.4.1
- Locust command line that you ran: locust --headless -u 1 -r 1 -t 100s --csv-full-history --csv test --only-summary
- Locust file contents (anonymized if necessary): see above.

**Comment by cyberw** [comment]
The "normal" way of doing this in locust would be to run multiple users instead of having one user do parallell requests. But if you need to do it this way, maybe you can increase urllib3:s pool size? Check the documentation for requests & urllib as this isnt really a locust thing.

Another option would be to spawn your own clients (locust.clients.HttpSession) and use them.

**Comment by alex4200** [comment]
I have seen [this documentation](https://urllib3.readthedocs.io/en/latest/advanced-usage.html#customizing-pool-behavior) on the pool size. So I just put these two line into the code `locustfile.py`?

```
import urllib3
http = urllib3.PoolManager(num_pools=50)
```

**Comment by alex4200** [comment]
No, this does not seem to work.

**Comment by cyberw** [comment]
Did you work it out? Sorry for not responding. This is not really a locust issue and requires more in depth knowledge of requests/urllib3 than I have...

**Comment by alex4200** [comment]
@cyberw No, I have not found a solution yet. Maybe I need to dive in to `requests/urllib` myself a bit more...

**Comment by aek** [comment]
Hi @alex4200 
Try it like this example(copy the `__init__` method to your Locust User)
```python
class CustomUser(HttpUser):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.client.mount('https://', HTTPAdapter(pool_maxsize=50))
        self.client.mount('http://', HTTPAdapter(pool_maxsize=50))

    def make_parallel_requests(self, requests):
        group = Group()
        uid = str(uuid.uuid4())[:8]
        print(f"{uid} Starting {len(requests)} requests in parallel.")
        for request in requests:
            if request["method"] == "GET":
                url = request["url"]
                headers = request["headers"]
                headers["Authorization"] = "bearer " + ACCESS_TOKEN
                group.spawn(lambda: self.client.get(url, headers=headers, verify=False))
        group.join()
        print(f"{uid} Requests done")

```

**Comment by alex4200** [comment]
@aek Thanks, that seems to work! 

And if someone else is interested in the solution, here the import of `HTTPAdapter`:

`from requests.adapters import HTTPAdapter`

**Comment by max-rocket-internet** [comment]
It would be cool we could set this on `HttpUser` instead of using `super` though, right?

**Comment by cyberw** [comment]
> It would be cool we could set this on `HttpUser` instead of using `super` though, right?

Sure! PR is welcome :)

**Comment by max-rocket-internet** [comment]
@aek can't wait to see your PR 🚀

**Comment by aek** [comment]
@max-rocket-internet I'm planning to test that idea by implementing a requests HTTPAdapter using pycurl.
That will be a proof of concept of this idea that could be backported to the rework of FastHTTPUser with the geventhttpclient converting it into a requests HTTPAdapter

**Comment by alex4200** [comment]
@aek Any news on this issue?

**Comment by aek** [comment]
Haven't had time to continue on this topic. I had researched the source codes and viability and had started the work but it's on hold yet due to others work loads that demand my time more. This is only a piece of the puzzle that I wanna do around this. I will get back here and to a couple of other locust issues with the results once it's done

**Comment by jalbinge** [comment]
customer_it_shop_roles = [
  "OG_ULM_PaymentsBI_PAY_BIPR_COMPLIANCE_entschluesselt@otto-peg.de",
  "OG_ULM_PaymentsBI_PAY_BIPR_OPS_B2C_entschluesselt@otto-peg.de",
"OG_ULM_PaymentsBI_PAY_BIPR_OPS_B2C_entschluesselt@otto-peg.de",
"OG_ULM_PaymentsBI_PAY_BIPR_LOOKER_DEVELOP@otto-peg.de"
......]

**Comment by jitendranair** [comment]
locust 2.14.2
python 3.9.2
Oracle-Linux 8.4

I saw this warning recently. 

Quoting the SO answer from here https://stackoverflow.com/a/66671026/2843837

**_For example, if your maxsize is 10 (the default when using urllib3 via requests), and you launch 50 requests in parallel, those 50 connections will be performed at once, and after completion only 10 will remain in the pool while 40 will be discarded (and issue that warning)._**

So I think it should not be a problem!
